### PR TITLE
SPARK-12925. Improve HiveInspectors.unwrap for StringObjectInspector.…

### PR DESCRIPTION
**Pull request by rajeshbalamohan**
Earlier fix did not copy the bytes and it is possible for higher level to reuse Text object. This was causing issues. Proposed fix now copies the bytes from Text. This still avoids the expensive encoding/decoding

### CODE REVIEW

**Comment by rxin** [comment]
@rajeshbalamohan in the future can you use the pull request template?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveInspectors.scala [hunk 1]
```diff
@@ -320,9 +320,11 @@ private[hive] trait HiveInspectors {
       case hvoi: HiveCharObjectInspector =>
         UTF8String.fromString(hvoi.getPrimitiveJavaObject(data).getValue)
       case x: StringObjectInspector if x.preferWritable() =>
-        // Text is in UTF-8 already. No need to convert again via fromString
+        // Text is in UTF-8 already. No need to convert again via fromString. Copy bytes
         val wObj = x.getPrimitiveWritableObject(data)
-        UTF8String.fromBytes(wObj.getBytes, 0, wObj.getLength)
+        val result = new Array[Byte](wObj.getLength())
```
Comment (srowen) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveInspectors.scala]: You can clone arrays with `.clone()`. Better yet, `Text` provides a `copyBytes()` method to return a copy of the bytes anyway.

**Comment by rajeshbalamohan** [comment]
Thanks @srowen . Incorporated the changes. 

This was tested with HiveCompatibilitySuite, HiveQuerySuite. These tests ran fine in master branch without the changes as well. However, when tried with 1.6 branch, these test suites failed with the copy issues. Hence doing explicit bytes copy in master, so that this does not fail in future.

**Comment by srowen** [comment]
@rajeshbalamohan you're saying this fixes a problem in 1.6 only? Looks OK to me.

**Comment by srowen** [comment]
Jenkins test this please

**Comment by srowen** [comment]
LGTM

**Comment by srowen** [comment]
Merged to master, since this fixes https://github.com/apache/spark/pull/10848 also merged only to master. If there's some other issue with 1.6 we should talk about it separately

**Pull request closed without merging**